src/txrequest.h

```diff
@@ -0,0 +1,210 @@
+    //
+    // - COMPLETED announcements represent transactions that have been requested from a peer, and a NOTFOUND or a
+    //   transaction was received in response (valid or not), or they timed out. They're only kept around to
+    //   prevent requesting them again. If only COMPLETED announcements for a given txhash remain (so no CANDIDATE
+    //   or REQUESTED ones), all of them are deleted (this is an invariant, and maintained by all operations below).
+    //
+    // The operations below manipulate the data structure.
+
+    /** Adds a new CANDIDATE announcement.
+     *
+     * Does nothing if one already exists for that (txhash, peer) combination (whether it's CANDIDATE, REQUESTED, or
+     * COMPLETED). Note that the txid/wtxid property is ignored for determining uniqueness, so if an announcement
+     * is added for a wtxid H, while one for txid H from the same peer already exists, it will be ignored. This is
+     * harmless as the txhashes being equal implies it is a non-segwit transaction, so it doesn't matter how it is
+     * fetched. The new announcement is given the specified preferred and reqtime values, and takes its is_wtxid
+     * from the specified gtxid.
+     */
+    void ReceivedInv(NodeId peer, const GenTxid& gtxid, bool preferred,
+        std::chrono::microseconds reqtime);
+
+    /** Deletes all announcements for a given peer.
+     *
+     * It should be called when a peer goes offline.
+     */
+    void DisconnectedPeer(NodeId peer);
+
+    /** Deletes all announcements for a given txhash (both txid and wtxid ones).
+     *
+     * This should be called when a transaction is no longer needed. The caller should ensure that new announcements
+     * for the same txhash will not trigger new ReceivedInv calls, at least in the short term after this call.
+     */
+    void ForgetTxHash(const uint256& txhash);
+
+    /** Find the txids to request now from peer.
+     *
+     * It does the following:
+     *  - Convert all REQUESTED announcements (for all txhashes/peers) with (expiry <= now) to COMPLETED ones.
+     *    These are returned in expired, if non-nullptr.
+     *  - Requestable announcements are selected: CANDIDATE announcements from the specified peer with
+     *    (reqtime <= now) for which no existing REQUESTED announcement with the same txhash from a different peer
+     *    exists, and for which the specified peer is the best choice among all (reqtime <= now) CANDIDATE
+     *    announcements with the same txhash (subject to preferredness rules, and tiebreaking using a deterministic
+     *    salted hash of peer and txhash).
+     *  - The selected announcements are converted to GenTxids using their is_wtxid flag, and returned in
+     *    announcement order (even if multiple were added at the same time).
+     *
+     * 'now' cannot go backwards. In every call, the highest value for 'now' ever passed is used, rather than the
```

Comment:
If the clock keeps going back-and-forth but without never crossing announcements' `m_time`, _DELAYED are going to stale forever ? Maybe worth to document to be sure we have all the same understanding of the new behavior.

Reply:
That seems absolutely trivial. If your clock makes no long-term progress then obviously no time-based events can be guaranteed to happen.

Reply:
This seems like it has slightly worse behaviour if time jumps backwards significantly (by more than the request/expiry delays). In that case:

 * any txs that are currently REQUESTED and get a NOTFOUND will result in all the CANDIDATE_READY alternatives being immediately requested (expiry will be set to the past so will expire immediately)
 * any new txs that come in will get immediately promoted from DELAYED to READY/BEST (reqtime in the past) and then be REQUESTED immediately, and then be immediately expired so any additional announcements that come in prior to the tx being delivered will also result in a request, etc.

I think if a bunch of systems have time problems simultaneously (bad daylight savings transition or ntp problems?) that could trigger some network spikes as many nodes request txs from many other nodes.

Making it the caller's responsibility to use a steady clock seems better than having txrequest try to fake it, IMO anyway.

Switching from `chrono::duration` to a `chrono::time_point` would allow you to test for `tp::clock::is_steady` and avoid actually doing any of the time-goes-backward code paths when you're using a clock where that's impossible, and also call `tp::clock::now()` directly rather than passing it in as a parameter (in which case the caller could cache old values and do them out of order etc to introduce bugs).

Would deferring the number-go-up requirement for time to a later PR make sense?

Reply:
Ok, removed the last commit, and moved it to https://github.com/sipa/bitcoin/commits/202010_txrequest_rand_wtxid_monotime if we want it still.